internal/core/writer/dimensions/client.go

```diff
@@ -0,0 +1,305 @@
+	PropertyFilterSet *propfilters.FilterSet
+	// For easier unit testing
+	now        func() time.Time
+	logUpdates bool
+
+	DimensionsCurrentlyDelayed int64
+	TotalDimensionsDropped     int64
+	// The number of dimension updates that happened to the same dimension
+	// within sendDelay.
+	TotalFlappyUpdates int64
+}
+
+type queuedDimension struct {
+	*types.Dimension
+	TimeToSend time.Time
+}
+
+// NewDimensionClient returns a new client
+func NewDimensionClient(ctx context.Context, conf *config.WriterConfig) (*DimensionClient, error) {
+	history, err := lru.New(int(conf.PropertiesHistorySize))
+	if err != nil {
+		panic("could not create properties history cache: " + err.Error())
+	}
+
+	propFilters, err := conf.PropertyFilters()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   5 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:        int(conf.PropertiesMaxRequests),
+			MaxIdleConnsPerHost: int(conf.PropertiesMaxRequests),
+			IdleConnTimeout:     30 * time.Second,
+			TLSHandshakeTimeout: 10 * time.Second,
+		},
+	}
+	sender := newReqSender(ctx, client, conf.PropertiesMaxRequests)
+
+	return &DimensionClient{
+		ctx:               ctx,
+		Token:             conf.SignalFxAccessToken,
+		APIURL:            conf.ParsedAPIURL(),
+		sendDelay:         time.Duration(conf.PropertiesSendDelaySeconds) * time.Second,
+		history:           history,
+		delayedSet:        make(map[types.DimensionKey]*types.Dimension),
+		delayedQueue:      make(chan *queuedDimension, conf.PropertiesMaxBuffered),
+		requestSender:     sender,
+		client:            client,
+		PropertyFilterSet: propFilters,
+		now:               time.Now,
+		logUpdates:        conf.LogDimensionUpdates,
+	}, nil
+}
+
+// Start the client's processing queue
+func (dc *DimensionClient) Start() {
+	go dc.processQueue()
+}
+
+// AcceptDimension to be sent to the API.  This will return fairly quickly and
+// won't block.  If the buffer is full, the dim update will be dropped.
+func (dc *DimensionClient) AcceptDimension(dim *types.Dimension) error {
+	filteredDim := &(*dim)
+
+	filteredDim = dc.PropertyFilterSet.FilterDimension(filteredDim)
+
+	dc.Lock()
+	defer dc.Unlock()
+
+	if delayedDim := dc.delayedSet[filteredDim.Key()]; delayedDim != nil {
+		dc.TotalFlappyUpdates++
+
+		// Don't further delay it if it gets updated so that we are always
+		// guaranteed to update a dimension at least some times, even if it
+		// continually gets updated more frequently than `sendDelay` seconds
+		// (which should be dealt with somewhere else).
+
+		// If the dim is a merge request, then update the existing one in
+		// place, otherwise replace it.
+		if delayedDim.MergeIntoExisting {
+			delayedDim.Properties = utils.MergeStringMaps(delayedDim.Properties, filteredDim.Properties)
```

Comment:
I don't think so because the filtering always happens before it gets to this point in the code.

Reply:
Do you see anything specific that is concerning?

Reply:
I was wondering if the dimensions set cache could end up with stale properties. But I don't think that's the case because it's guaranteed that we have a clean cache whenever the property filters are updated.